### DISCUSSION TITLE
SixDOFConstraint Confusion

### DISCUSSION THREAD

**Discussion by HunterBarclay** [jrouwe/JoltPhysics#1188] [Q&A]
Hello, I'm currently trying to create a ball socket constraint. Forgive me if this isn't the best question but in what order are the rotational rotational degrees of freedom applied? If I were to say lock out the Z axis, restrict the X axis to -45 to 45 degrees, and leave the Y free, would that create a cone area of movement around the Y axis? Do the axes that is restricts by orient themselves relative to one of the bodies?

**Comment by jrouwe** [comment]
Hello,

I'm assuming you're using a `SixDOFConstraint`? This constraint uses [swing twist decomposition](http://www.mathforgameprogrammers.com/gdc2016/GDC2016_VanDenBergen_Gino_RotationalJointLimits.pdf). The twist is around X and is applied first, the swing is around Y and Z and is applied next. The YZ rotation is encoded in a quaternion, so there's not really an order to it.

If you want to leave one axis free, make it the X axis (and update the `SixDOFConstraintSettings::mAxisX1` so that it points to where your Y axis used to point to).

The best way of playing around with this is to start the samples app, hit ESC and select 'Select Test' / 'Constraint' / 'Six DOF Test'. Press the C (= show constraints) and L (= show limits) buttons to visualize the constraints. Then select 'Test Settings' / 'Configuration Settings (Limits)' and e.g. set the settings according to what you proposed above:

![image](https://github.com/user-attachments/assets/48241ddc-c077-475d-8b21-dea81ec17e18)

And select 'Accept Changes'.

Then hit P to unpause and point the center of the screen to the lower body and pull it by holding space bar.
